docs/metrics-data-model.md

```diff
@@ -0,0 +1,177 @@
+# k6 Time Series
+
+We would migrate to a time series data model for getting the benefits of aggregated data so just allocating samples one time and pass around the aggregated time series value (or values - e.g. in case of Trend). The Outputs will switch from fetching the buffered samples to fetching a snapshot of the time series (the same concept adopted by Prometheus' scraping operation).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* TimeSeries (metric, tagset)
+* Sample (metric, tagset, timestamp, value)
+* TimeSeries Sample (timeseries ref, timestamp, value)
+
+```go
+// This is the current k6 Metric polished from the Thresholds stuff.
+type Metric struct {
+	Name     string     `json:"name"`
+	Type     MetricType `json:"type"`
+	Contains ValueType  `json:"contains"`
+}
+
+type TimeSeries struct {
+    ID uin64 // hash(metric+tags)
```

Comment:
My assumption: the set of hashes must be unique per test run.

In case of a distributed test (e.g. k6-operator) with the following setup:

Instances: 1000
Unique time series generated from the test: 10k
Total time series = 10M

Where the numbers are nearly unrealistic for our use cases, right?

But even in this case, it seems we are far from the probability for a 50% collision which is around 4 billion ($2^{64/2}$) of generated hashes required. At least this theoretically.

The Go's `hash/maphash` implementation seems not good enough for our case, because it doesn't guarantee the same output across different processes, I don't know if we really require it, but it could be a nice to have for supporting future features. In any case, other solutions like the [xxhash](https://github.com/Cyan4973/xxHash) seem good enough to me.

Reply:
Hmm I've only heard of xxhash, I haven't actually used it, so I can't comment on it too much, but seems like a nice hashing algorithm. But even if we assume it's perfect, I was a bit skeptical about this assertion: 

> But even in this case, it seems we are far from the probability for a 50% collision which is around 4 billion ($2^{64/2}$) of generated hashes required. At least this theoretically.

Because of the ["birthday problem"](https://en.wikipedia.org/wiki/Birthday_problem), collision calculations are somewhat unintuitive. I tried to answer the question "What are the chances that there are collisions with a perfect 64-bit hash in even one out of 100000000 (100m) distinct load tests, each with 100000 (100k) unique time series?", which is probably way beyond even the most realistic worst-case scenario? And I _think_ the I math is somewhat like `1-(e^(-(100000^2)/(2*(2^64))))^100000000`, based on what I remember from statistics and [this article](https://preshing.com/20110504/hash-collision-probabilities/). The result is 2% chance, which is probably good enough for us :sweat_smile: Still, I wonder what hash function and length Prometheus and other similar tools use? :thinking: